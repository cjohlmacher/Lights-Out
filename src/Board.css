.Board {
  margin: 5px auto;
  border: 2px solid black;
}