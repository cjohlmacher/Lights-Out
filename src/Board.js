import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows=5, ncols=5, chanceLightStartsOn=0.5 }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];
    let rowCount = 0;
    while (rowCount < nrows) {
      let lightRow = Array.from({length: ncols});
      lightRow = lightRow.map(cell => true);
      initialBoard.push(lightRow);
      rowCount += 1;
    };
    initialBoard = randomizeLights(initialBoard);
    return initialBoard;
  };

  /**  Given a board configuration, return a board with flips around cell at coord 'Y-X' **/
  function flipCellsOfBoard(preboard,coord) {
    const [y, x] = coord.split("-").map(Number);
    
    const flipCell = (y, x, boardCopy) => {
      // if this coord is actually on board, flip it
      if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
        boardCopy[y][x] = !boardCopy[y][x];
      }
    };

    const newBoard = preboard.map(row => row.map(cell => cell));
    const cellsToFlip = [[y,x],[y-1,x],[y+1,x],[y,x-1],[y,x+1]];
    cellsToFlip.forEach(([y,x]) => flipCell(y,x,newBoard));
    return newBoard;
  }

  /** Randomize lit/unlit lights **/
  function randomizeLights(initialBoard) {
    const targetLit = chanceLightStartsOn*nrows*ncols;
    let litCount = nrows*ncols;
    while (litCount  > targetLit) {
      const [randomY, randomX] = [Math.floor(Math.random()*nrows), Math.floor(Math.random()*ncols)];
      initialBoard = flipCellsOfBoard(initialBoard,`${randomY}-${randomX}`);
      litCount = 0;
      for (let row of initialBoard) {
        for (let cell of row) {
          if (cell) {
            litCount += 1
          }
        }
      };
    };
    return initialBoard
  };

  /** Determine if board is won */
  function hasWon() {
    for (let row of board) {
      for (let cell of row) {
        if (!cell) {
          return false
        }
      }
    };
    return true
  };

  /** Set board in state based on flipping cells around coord 'Y-X' **/
  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      return flipCellsOfBoard(oldBoard,coord);
    });
  };

  let rowNum=-1;
  let columnNum=-1;
  
  /**  Create light grid HTML elements **/
  const lightGrid = (
    <table>
      <tbody>
        {board.map(row => {
          rowNum += 1;
          columnNum = -1;
          return (
            <tr key={rowNum}>
              {
              row.map(cell => {
                columnNum += 1;
                const key = `${rowNum}-${columnNum}`;
                return <Cell key={key} flipCellsAroundMe={() => flipCellsAround(key)} isLit={cell} />
                })
              }
            </tr>
          )
        })}
      </tbody>
    </table>
  );

  return (
    <>
      {hasWon() ? <h1>You Win!</h1> : lightGrid}
    </>
  )
}

export default Board;
