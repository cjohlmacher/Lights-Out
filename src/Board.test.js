import { fireEvent, getByText, render } from '@testing-library/react';
import Board from './Board';

beforeEach(function() {
    jest
        .spyOn(Math, "random")
        .mockReturnValueOnce(0.01)
        .mockReturnValueOnce(0.01)
        .mockReturnValueOnce(0.875)
        .mockReturnValueOnce(0.875)
        .mockReturnValueOnce(0.5)
        .mockReturnValueOnce(0.5)
        .mockReturnValueOnce(0.725)
        .mockReturnValueOnce(0.725)
        .mockReturnValueOnce(0.375)
        .mockReturnValueOnce(0.375)
        .mockReturnValueOnce(0.125)
        .mockReturnValueOnce(0.125)
        .mockReturnValueOnce(0.625)
        .mockReturnValueOnce(0.625)
        .mockReturnValueOnce(0.125)
        .mockReturnValueOnce(0.125)
        .mockReturnValueOnce(0.5)
        .mockReturnValueOnce(0.5);
});

it('renders without crashing', () => {
    render(<Board />);
});

it('matches snapshot with 3x3 grid', () => {
    const {asFragment} = render(<Board nrows={3} ncols={3}/>);
    expect(asFragment()).toMatchSnapshot();
});

it('flips cell on click', () => {
    const { getByRole, debug } = render(<Board nrows={3} ncols={3}/>);
    const board = getByRole('table');
    const tbody = board.children[0];
    const grid = [
        [tbody.children[0].children[0],tbody.children[0].children[1],tbody.children[0].children[2]],
        [tbody.children[1].children[0],tbody.children[1].children[1],tbody.children[1].children[2]],
        [tbody.children[2].children[0],tbody.children[2].children[1],tbody.children[2].children[2]]
    ];
    const cell1 = grid[0][0];
    fireEvent.click(cell1);
    expect(cell1.className).toBe('Cell Cell-lit');
    expect(grid[0][1].className).toBe('Cell Cell-lit');
    expect(grid[0][2].className).toBe('Cell Cell-lit');
    expect(grid[1][0].className).toBe('Cell Cell-lit');
    expect(grid[1][1].className).toBe('Cell Cell-lit');
    expect(grid[1][2].className).toBe('Cell ');
    expect(grid[2][0].className).toBe('Cell Cell-lit');
    expect(grid[2][1].className).toBe('Cell ');
    expect(grid[2][2].className).toBe('Cell ');
});

it('shows win', () => {
    const { getByRole, getByText, debug } = render(<Board nrows={3} ncols={3}/>);
    const board = getByRole('table');
    const tbody = board.children[0];
    const grid = [
        [tbody.children[0].children[0],tbody.children[0].children[1],tbody.children[0].children[2]],
        [tbody.children[1].children[0],tbody.children[1].children[1],tbody.children[1].children[2]],
        [tbody.children[2].children[0],tbody.children[2].children[1],tbody.children[2].children[2]]
    ];
    const cell1 = grid[0][0];
    const cell2 = grid[2][2];
    fireEvent.click(cell1);
    fireEvent.click(cell2);
    const heading = getByText('You Win!');
    expect(heading).toBeInTheDocument();
    expect(board).not.toBeInTheDocument();
});