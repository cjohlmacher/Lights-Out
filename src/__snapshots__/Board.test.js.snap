// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot with 3x3 grid 1`] = `
<DocumentFragment>
  <table
    class="Board"
  >
    <tbody>
      <tr>
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
      </tr>
      <tr>
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
